src/librustc/ty/subst.rs

```diff
@@ -39,6 +39,20 @@ const TAG_MASK: usize = 0b11;
 const TYPE_TAG: usize = 0b00;
 const REGION_TAG: usize = 0b01;
 
+pub enum GenericParameterKind<'tcx> {
```

Comment:
Do we really need this? I know `Kind` is a bit awkward to work with... but if you an `UnpackedKind` it should maybe be a separate PR? Because you should *remove* `as_{lifetime,type}` and *force* everything to go through `UnpackedKind`.

Reply:
It simplifies one loop so far. I'll rename it now and think about whether it's worth committing to or not.